### PR TITLE
[CALCITE-6312] Add LOG function (enabled in PostgreSQL library)

### PR DESCRIPTION
**Pull request by caicancai**
https://issues.apache.org/jira/browse/CALCITE-6312
https://issues.apache.org/jira/browse/CALCITE-6456

### CODE REVIEW

**Comment by caicancai** [comment]
In order to prevent LogImplementor from making the code redundant, I added SqlLibrary for identification

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java [hunk 1]
```diff
@@ -4202,67 +4205,43 @@ private static class LogicalNotImplementor extends AbstractRexCallImplementor {
...
+        List<Expression> argValueList, int flag) {
       Expression operand0 = argValueList.get(0);
       final Expressions.FluentList<Expression> list = Expressions.list(operand0);
       switch (call.getOperator().getName()) {
       case "LOG":
         if (argValueList.size() == 2) {
-          return list.append(argValueList.get(1)).append(Expressions.constant(1));
+          return list.append(argValueList.get(1)).append(Expressions.constant(flag));
+        }
+        if (flag == 2) {
```
Comment (YiwenWu) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:4234]: Whether to consider BigQuery case, flag=0 ?

Comment (caicancai) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:4234]: Thanks for the review, I will improve the annotations to explain it later.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 2]
```diff
@@ -7045,6 +7049,33 @@ void checkRegexpExtract(SqlOperatorFixture f0, FunctionAlias functionAlias) {
...
+  @Test void testPostgresLogFunc() {
+    final SqlOperatorFixture f0 = fixture()
+        .setFor(SqlLibraryOperators.LOG_POSTGRES, VmName.EXPAND);
+    f0.checkFails("^log(100, 10)^",
+        "No match found for function signature LOG\\(<NUMERIC>, <NUMERIC>\\)", false);
+    final SqlOperatorFixture f = f0.withLibrary(SqlLibrary.POSTGRESQL);
+    f.checkScalarApprox("log(10, 10)", "DOUBLE NOT NULL",
+        isWithin(1.0, 0.000001));
+    f.checkScalarApprox("log(64, 8)", "DOUBLE NOT NULL",
+        isWithin(2.0, 0.000001));
```
[the remark this answers is not in the document]
Comment (caicancai) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java]: done

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 3]
```diff
@@ -7045,6 +7049,33 @@ void checkRegexpExtract(SqlOperatorFixture f0, FunctionAlias functionAlias) {
     f0.forEachLibrary(list(SqlLibrary.MYSQL, SqlLibrary.SPARK), consumer);
   }
 
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-6321">[CALCITE-6321]
```
Comment (YiwenWu) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:7053]: Jira link error

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 4]
```diff
@@ -7045,6 +7049,33 @@ void checkRegexpExtract(SqlOperatorFixture f0, FunctionAlias functionAlias) {
...
+   * Add LOG function (enabled in PostgreSQL library)</a>. */
+  @Test void testPostgresLogFunc() {
+    final SqlOperatorFixture f0 = fixture()
+        .setFor(SqlLibraryOperators.LOG_POSTGRES, VmName.EXPAND);
+    f0.checkFails("^log(100, 10)^",
+        "No match found for function signature LOG\\(<NUMERIC>, <NUMERIC>\\)", false);
+    final SqlOperatorFixture f = f0.withLibrary(SqlLibrary.POSTGRESQL);
+    f.checkScalar("log(10, 10)", 1.0,
+        "DOUBLE NOT NULL");
+    f.checkScalar("log(64, 8)", 2.0,
```
Comment (YiwenWu) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:7063]: LOG(base, value), the result of log(64, 8) is not 2.0

**Review by YiwenWu — CHANGES_REQUESTED** [review]

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 4 above)
Comment (caicancai) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:7063]: ohhh, seems to be a bug in sqlfunctions, I open a new jira case, bigquery is log(64, 8) == 2?
I overlooked the problem, and I'm sorry

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java [hunk 5]
```diff
@@ -4202,67 +4205,43 @@ private static class LogicalNotImplementor extends AbstractRexCallImplementor {
...
-    }
-
-    @Override Expression implementSafe(final RexToLixTranslator translator,
-        final RexCall call, final List<Expression> argValueList) {
-      return Expressions.call(BuiltInMethod.LOG.method, args(call, argValueList));
-    }
 
     private static List<Expression> args(RexCall call,
-        List<Expression> argValueList) {
+        List<Expression> argValueList, int flag) {
```
Comment (XuQianJin-Stars) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:4226]: The args has been changed and the default value of flag is 0. Have compatibility issues been considered here?

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java (hunk 1 above)
Comment (caicancai) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java]: fix

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java [hunk 6]
```diff
@@ -4202,67 +4205,54 @@ private static class LogicalNotImplementor extends AbstractRexCallImplementor {
...
-      return Expressions.call(BuiltInMethod.LOG.method, args(call, argValueList));
-    }
 
     private static List<Expression> args(RexCall call,
-        List<Expression> argValueList) {
-      Expression operand0 = argValueList.get(0);
+        List<Expression> argValueList, int flag) {
+      Expression operand0;
+      Expression operand1;
+      if (flag == 0 && argValueList.size() == 2) {
```
Comment (caicancai) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:4229]: I don't know if this part of the code can be optimized, but it seems to be the only way to match it

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 4 above)
Comment (caicancai) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java]: fix

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java (hunk 5 above)
Comment (caicancai) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:4226]: done

**Review by YiwenWu — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java [hunk 7]
```diff
@@ -4202,67 +4205,55 @@ private static class LogicalNotImplementor extends AbstractRexCallImplementor {
...
+        operand0 = argValueList.get(0);
+        operand1 = argValueList.get(1);
+      } else if ((flag == 1 || flag == 2) && argValueList.size() == 2) {
```
[the remark this answers is not in the document]
Comment (YiwenWu) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:4232]:
```
  private static class LogImplementor extends AbstractRexCallImplementor {

    @Override Expression implementSafe(final RexToLixTranslator translator,
        final RexCall call, final List<Expression> argValueList) {
      return Expressions.call(BuiltInMethod.LOG.method, args(call, argValueList, library));
    }

    private static List<Expression> args(RexCall call,
        List<Expression> argValueList, SqlLibrary library) {
      Pair<Expression, Expression> baseAndValuePair;
      if (argValueList.size() == 1) {
        ConstantExpression defaultBase = library == SqlLibrary.POSTGRESQL
            ? Expressions.constant(BigDecimal.TEN) : Expressions.constant(Math.exp(1));
        baseAndValuePair = Pair.of(defaultBase, argValueList.get(0));
      } else {
        boolean isBaseFirst = library == SqlLibrary.MYSQL || library == SqlLibrary.POSTGRESQL;
        baseAndValuePair = isBaseFirst ? Pair.of(argValueList.get(0), argValueList.get(1))
            : Pair.of(argValueList.get(1), argValueList.get(0));
      }
      Expression base = baseAndValuePair.left;
      Expression value = baseAndValuePair.right;
      final Expressions.FluentList<Expression> list = Expressions.list();
      switch (call.getOperator().getName()) {
      case "LOG":
        return list.append(value).append(base);
      case "LN":
        return list.append(value).append(Expressions.constant(Math.exp(1)));
      case "LOG2":
        return list.append(value).append(Expressions.constant(2));
      case "LOG10":
        return list.append(value).append(Expressions.constant(BigDecimal.TEN));
      default:
        throw new AssertionError("Operator not found: " + call.getOperator());
      }
    }
  }
```

[the remark this answers is not in the document]
Comment (caicancai) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java]: thanks

[the remark this answers is not in the document]
Comment (caicancai) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java]: Thank you for your code. I will optimize this part based on your code. Thank you very much.

[the remark this answers is not in the document]
Comment (caicancai) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java]: You also need to consider the flag, because mysql's log(0) will return null, but bigqeury and pg will not.

**Comment by caicancai** [comment]
@XuQianJin-Stars @YiwenWu I optimized the implementation of the adapted log function. Can you review it for me when you have time? Thank you @YiwenWu for your suggestions.

**Comment by caicancai** [comment]
@YiwenWu @XuQianJin-Stars @NobiGo PATL,  thanks

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java [hunk 8]
```diff
@@ -2237,13 +2237,21 @@ private static RelDataType deriveTypeMapFromEntries(SqlOperatorBinding opBinding
...
 
-  /** The "LOG(numeric, numeric1)" function. Returns the base numeric1 logarithm of numeric. */
+  /** The "LOG(numeric[,numeric1])" function. Returns the base numeric1 logarithm of numeric. */
   @LibraryOperator(libraries = {MYSQL, SPARK})
   public static final SqlFunction LOG_MYSQL =
       SqlBasicFunction.create(SqlKind.LOG,
           ReturnTypes.DOUBLE_FORCE_NULLABLE,
           OperandTypes.NUMERIC_OPTIONAL_NUMERIC);
 
+  /** The "LOG(numeric[,numeric1])" function. Returns the base numeric1 logarithm of numeric. */
```
Comment (YiwenWu) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java:2247]: The description is a bit confusing.
'Returns the logarithm of numeric1 using numeric as the base'?

File: site/_docs/reference.md [hunk 9]
```diff
@@ -2805,8 +2805,9 @@ In the following:
 | f s | LEN(string)                                  | Equivalent to `CHAR_LENGTH(string)`
 | b f s | LENGTH(string)                             | Equivalent to `CHAR_LENGTH(string)`
 | h s | LEVENSHTEIN(string1, string2)                | Returns the Levenshtein distance between *string1* and *string2*
-| b | LOG(numeric1 [, numeric2 ])                    | Returns the logarithm of *numeric1* to base *numeric2*, or base e if *numeric2* is not present
+| b | LOG(numeric1 [, numeric2 ])                    | Returns the logarithm of *numeric1* to base *numeric2*, or base e if *numeric2* is not present if 1-arg version uses base e(ln)
 | m s | LOG(numeric1 [, numeric2 ])                  | Returns the logarithm of *numeric1* to base *numeric2*, or base e if *numeric2* is not present, or null if *numeric1* is 0 or negative
+| p | LOG(numeric1 [, numeric2 ])                    | Returns the logarithm of *numeric1* to base *numeric2*, or base e if *numeric2* is not present if 1-arg version uses base 10
```
Comment (YiwenWu) [inline site/_docs/reference.md:2810]: "Returns the logarithm of *numeric1* to base *numeric2*",  → "numeric2" is base

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 8 above)
Comment (caicancai) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java]: fix

**Review by YiwenWu — APPROVED** [review]
LGTM

File: core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java (hunk 8 above)
Comment (caicancai) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlLibraryOperators.java]: I modified it according to the official MySQL

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java [hunk 10]
```diff
@@ -4202,67 +4206,46 @@ private static class LogicalNotImplementor extends AbstractRexCallImplementor {
...
+      operand1 = operands.right;
       final Expressions.FluentList<Expression> list = Expressions.list(operand0);
       switch (call.getOperator().getName()) {
       case "LOG":
-        if (argValueList.size() == 2) {
-          return list.append(argValueList.get(1)).append(Expressions.constant(1));
-        }
-        // fall through
+        return list.append(operand1).append(Expressions.constant(flag));
+          // fail through
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:4241]: it is 'fall', not 'fail'

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 11]
```diff
@@ -2788,38 +2788,50 @@ public static double power(BigDecimal b0, BigDecimal b1) {
 
   // LN, LOG, LOG10, LOG2
 
-  /** SQL {@code LOG(number, number2)} function applied to double values. */
-  public static @Nullable Double log(double number, double number2, int nullFlag) {
-    if (nullFlag == 1 && number <= 0) {
+  /** SQL {@code LOG(number, base)} function applied to double values. */
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:2791]: The meaning of 'flag' is very hard to follow from the call to the callee.
I think that a better way is to really have different functions, corresponding to the 3 different versions of 'flag'. Each function can have a nice descriptive name, like log_base_value

**Review by mihaibudiu — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java (hunk 11 above)
Comment (caicancai) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:2791]: I'm worried that this makes the code redundant, so I thought I could add a comment to make this clear.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java (hunk 10 above)
Comment (caicancai) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:4241]: fall through will cause cli to fail, I don't know why, but fail through will not

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:4241]: this actually does not fall through, so you should just delete the comment.

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java [hunk 12]
```diff
@@ -2785,41 +2785,69 @@ public static double power(BigDecimal b0, BigDecimal b1) {
...
-  public static @Nullable Double log(double number, double number2, int nullFlag) {
-    if (nullFlag == 1 && number <= 0) {
+  /**
+   * SQL {@code LOG(number, base)} function applied to double values.
+   *
+   * @param dialectflag This parameter is used to specify the SQL dialect for the LOG function.
+   * 0 represents MySQL dialect, 1 represents BigQuery dialect, and 2 represents PostgreSQL dialect.
+   */
+  public static @Nullable Double log(double number, double base, int dialectflag) {
+    if (dialectflag == 0 && number <= 0) {
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:2797]: you only compare dialectflag with 0.
So this flag should be named something like "nonPositiveIsNull".
The JavaDoc should say: "if true return null for non-positive values"

**Review by mihaibudiu — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java (hunk 12 above)
Comment (caicancai) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java:2797]: yes, you are right

Comment (caicancai) [reply core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java]: done, thank you

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java [hunk 13]
```diff
@@ -4202,67 +4206,48 @@ private static class LogicalNotImplementor extends AbstractRexCallImplementor {
...
+    private final SqlLibrary library;
+    LogImplementor(SqlLibrary library) {
       super("log", NullPolicy.STRICT, true);
+      this.library = library;
     }
 
     @Override Expression implementSafe(final RexToLixTranslator translator,
         final RexCall call, final List<Expression> argValueList) {
-      return Expressions.call(BuiltInMethod.LOG.method, args(call, argValueList));
+      boolean nonPositiveIsNull = library == SqlLibrary.MYSQL ? true : false;
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:4217]: this can be simplified by removing the ? and everything after it.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 14]
```diff
@@ -6965,8 +6965,16 @@ void checkRegexpExtract(SqlOperatorFixture f0, FunctionAlias functionAlias) {
         isWithin(9.0, 0.000001));
     f.checkScalarApprox("log(cast(10e-3 as real), 10)", "DOUBLE NOT NULL",
         isWithin(-2.0, 0.000001));
+    f.checkScalarApprox("log(10)", "DOUBLE NOT NULL",
+        isWithin(2.302585092994046, 0.000001));
     f.checkNull("log(cast(null as real), 10)");
     f.checkNull("log(10, cast(null as real))");
+    f.checkFails("log(0)",
+        "Cannot take logarithm of zero or negative number", true);
+    f.checkFails("log(0, 64)",
```
Comment (mihaibudiu) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:6974]: The error message could probably be more descriptive, showing the value which is illegal.
Also, in one of these two cases it's not the number which is illegal, but the base.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 15]
```diff
@@ -7017,24 +7025,26 @@ void checkRegexpExtract(SqlOperatorFixture f0, FunctionAlias functionAlias) {
...
+      f.checkScalarApprox("log(3,27)", "DOUBLE",
           isWithin(3.0, 0.000001));
-      f.checkScalarApprox("log(100, 10)", "DOUBLE",
-          isWithin(2.0, 0.000001));
       f.checkScalarApprox("log(10, 100)", "DOUBLE",
+          isWithin(2.0, 0.000001));
+      f.checkScalarApprox("log(100, 10)", "DOUBLE",
           isWithin(0.5, 0.000001));
-      f.checkScalarApprox("log(cast(10e6 as double), 10)", "DOUBLE",
+      f.checkScalarApprox("log(10, cast(10e6 as double))", "DOUBLE",
```
Comment (mihaibudiu) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:7036]: I think these tests would be easier to read if you wrote 1e7.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 16]
```diff
@@ -7017,24 +7025,26 @@ void checkRegexpExtract(SqlOperatorFixture f0, FunctionAlias functionAlias) {
...
           isWithin(7.0, 0.000001));
-      f.checkScalarApprox("log(cast(10e8 as float), 10)", "DOUBLE",
+      f.checkScalarApprox("log(10, cast(10e8 as float))", "DOUBLE",
           isWithin(9.0, 0.000001));
-      f.checkScalarApprox("log(cast(10e-3 as real), 10)", "DOUBLE",
+      f.checkScalarApprox("log(10, cast(10e-3 as real))", "DOUBLE",
```
Comment (mihaibudiu) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:7040]: maybe the fact that the log of a REAL is a DOUBLE should be documented

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java [hunk 17]
```diff
@@ -4202,67 +4206,48 @@ private static class LogicalNotImplementor extends AbstractRexCallImplementor {
...
-          return list.append(argValueList.get(1)).append(Expressions.constant(0));
-        }
-        // fall through
-      case "LN":
-        return list.append(Expressions.constant(Math.exp(1))).append(Expressions.constant(0));
-      case "LOG10":
-        return list.append(Expressions.constant(BigDecimal.TEN)).append(Expressions.constant(0));
-      default:
-        throw new AssertionError("Operator not found: " + call.getOperator());
+        List<Expression> argValueList, SqlLibrary library, boolean nonPositiveIsNull) {
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:4223]: I would add JavaDoc here.
Since you are passing the library, you  as an argument, you can also compute nonPositiveIsNull in this function instead of passing it as an argument. I think this would be more consistent with the way you look at the other libraries.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 14 above)
Comment (caicancai) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:6974]: In fact, in LOG(numeric1 [, base ]) these two parameters cannot both be 0 or negative, otherwise errors will be reported in postgres and bigqeury.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java (hunk 17 above)
Comment (caicancai) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java]: done

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java (hunk 13 above)
Comment (caicancai) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java]: done

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 15 above)
Comment (caicancai) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java]: done

**Comment by caicancai** [comment]
@mihaibudiu PATL, thank you

File: core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java [hunk 18]
```diff
@@ -4202,67 +4206,65 @@ private static class LogicalNotImplementor extends AbstractRexCallImplementor {
...
         final RexCall call, final List<Expression> argValueList) {
-      return Expressions.call(BuiltInMethod.LOG.method, args(call, argValueList));
-    }
-
+      return Expressions.
+          call(BuiltInMethod.LOG.method, args(call, argValueList, library));
+    }
+
+    /**
+     * This method, `args`, is used to handle the arguments for logarithmic functions in SQL.
```
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java:4222]: what is "`args`" here?
If you want to document the parameters please use the proper JavaDoc format using `@parameter` to describe what each parameter does.
This is almost ready.

Comment (caicancai) [reply core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java]: done

**Review by mihaibudiu — APPROVED** [review]

**Comment by caicancai** [comment]
@mihaibudiu @YiwenWu @XuQianJin-Stars Thank you for your review
